Remove path strategy chooser from driver

peterschrammel opened the pull request:
Makes `parse_path_strategy_options` a function. A BMC-algorithm-specific member shouldn't be owned by *bmc_parse_optionst; only the chosen option should be validated and passed. 

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Make parse_path_strategy_options a function

Removes an unnecessary member from *bmc_parse_optionst.

To implement an incremental_goto_checker for --paths,
we have to encapsulate everything algorithm-specfic
inside the incremental_goto_checker.
We can only parse the option in the driver, but
cannot pass on the path_strategy_chooser instance.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eecdb00133ae25ec66c4fc14a60653a247d3291f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97178570

smowton commented:
`path_strategy_choosert` is actually completely stateless; instead of making one and passing it to do_language_agnostic_bmc, how about simply making them free functions (the map `strategies` should probably be a static data structure, not a class member)

peterschrammel commented:
Yes, right, @smowton. I was a bit lazy and wanted to avoid breaking TG and SS by changing the interface.

smowton commented:
I say break them, this is a pretty new interface, so the sooner it's made right the better! Plus I don't believe TG uses do-language-agnostic anyhow.

peterschrammel committed:
Refactor path strategy chooser into free functions

Since keeping an instance is not an option,
we should avoid having to create an instance on every
stateless function call.

peterschrammel commented:
@smowton, please have a look whether 903b059 is what you envisaged.

smowton commented:
Yes that looks about right

peterschrammel commented:
> Yes that looks about right

@smowton, can you approve it?

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 50a9ef15703e8b41ff35afb4682b5c5b0985c1fd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97562758
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.
